### DISCUSSION TITLE
ERROR -- : OpenTelemetry error: Unable to export 4 spans

### DISCUSSION THREAD

**Discussion by ruby4mag** [open-telemetry/opentelemetry-ruby#966] [Q&A] [answered]
Hello ,

I get the following error on rails console 

`E, [2021-09-29T21:34:04.182294 #54775] ERROR -- : OpenTelemetry error: Unable to export 4 spans`

and dont see traces getting exported to zipkins (deployed using the sample docker config in examples)

i have set the following

`export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318/v1/traces`

config/initializers/opentelemetry.rb

```
require 'opentelemetry/sdk'
require 'opentelemetry/exporter/otlp'
require 'opentelemetry/instrumentation/all'
OpenTelemetry::SDK.configure do |c|
  c.service_name = 'my-service'
  c.use_all() # enables all instrumentation!
end
```

and can see the otel exporter is listening on 4318. Any ideas what could be wrong?

Thanks,
-Magin

**Comment by arielvalentin** [comment, marked as answer]
It is difficult to say with the information you have provided. Does the data get to the collector at all? Does it go to Jaeger?

Could you share how you determined the collector was listening on `localhost:4318`? 

Try doing the following:

1. Set log level to debug [`export OTEL_LOG_LEVEL=debug`](https://github.com/open-telemetry/opentelemetry-ruby/blob/11c97228f35ce9338175e63de774a521d593c55a/api/lib/opentelemetry.rb#L31) 
2. Add a [`custom error handler` ](https://github.com/open-telemetry/opentelemetry-ruby/blob/11c97228f35ce9338175e63de774a521d593c55a/sdk/test/opentelemetry/sdk_test.rb#L20) to print out the stack trace or more details about the error.

**Reply by ruby4mag** [reply]
Hello Arielvalentin,

Thanks for the hint. enabling custom handler helped in identifying the issue. 

`DEBUG -- : Error extracting W3C baggage: undefined method `gsub' for nil:NilClass`

This was because of version opentelemetry-exporter-otlp  0.17.0 

I upgraded to version 0.20.1and error went off.
Another thing I had to do was setting the OTLP destination as below.

`ENV['OTEL_EXPORTER_OTLP_TRACES_ENDPOINT']='http://0.0.0.0:4318/v1/traces'`

After this trace exporting worked fine.

Thanks for all your help Arielvaletin !!

-Magin

**Reply by almeidaalex** [reply]
@ruby4mag Hello, I'm facing the same situation that you were in, but I'm a newbie on RoR and Otel.
Where did you set the endpoint? I set it as an env variable, following the https://opentelemetry.io/docs/ruby/getting_started/ but I'm still having the same error.

And how did you get the error details? Because I did this, but it isn't working well
```
OpenTelemetry.error_handler = lambda do |exception: nil, message: nil|
    received_exception = exception
    received_message = message

    puts received_message.to_s
    puts received_exception.to_s
end
```
Thank you

**Reply by arielvalentin** [reply]
It appears that something has changed. 

The example relies on the [latest dev image of the otel-collector](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/examples/otel-collector/.env#L1). It looks like something has changed from the time that I [copied over the configuration](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/examples/otel-collector/otel-collector-config.yaml#L16) because the `jaeger` exporter requires the `insecure` attribute to be nested under `tls`.  

It also relied on `docker-compose` using `network="host"` so it assumed the collector would listen on hostname `localhost` to resolve to `127.0.0.1`. I am going to make a few updates to the docs to reflect the most up to date configurations.

```
$> docker-compose up
# some output...  
Attaching to otel-collector_jaeger-all-in-one_1, otel-collector_zipkin-all-in-one_1, otel-collector_otel-collector_1
otel-collector_1     | Error: cannot load configuration: error reading exporters configuration for jaeger: 1 error(s) decoding:
otel-collector_1     | 
otel-collector_1     | * '' has invalid keys: insecure
otel-collector_1     | 2021/10/04 21:18:01 collector server run finished with error: cannot load configuration: error reading exporters configuration for jaeger: 1 error(s) decoding:
otel-collector_1     | 
otel-collector_1     | * '' has invalid keys: insecure
otel-collector_otel-collector_1 exited with code 1
```

Related: https://github.com/open-telemetry/opentelemetry-collector/issues/4092

**Reply by arielvalentin** [reply]
Thanks for reporting this problem I have a PR here. 

Please test out these changes whenever you are ready: https://github.com/open-telemetry/opentelemetry-ruby/pull/970

**Reply by almeidaalex** [reply]
Hi @arielvalentin thank you create the PR to correct this issue, but it hasn't been approved yet.
I've made two changes in order to collect data from the application to Collector.
1 - I changed the jaeger section from `otel-collector-config.yaml`  from
``` yaml
jaeger:
    endpoint: jaeger-all-in-one:14250
    insecure: true
```
to:
``` yaml
jaeger:
    endpoint: jaeger-all-in-one:14250
    tls:
      insecure: true
```
Is it a change from 0.36 version?

2 - Remove the path from the URL, I just left the host, as you made on PR, from `http://localhost:4318/v1/traces` to `http://localhost:4318/`

Now it's working, but it was tricky to find a solution. Is it possible to add on error, the reason why the spans cannot be sent?

Thank you so much!
